redis-test/test_strings.py

```diff
@@ -113,3 +113,23 @@ def test_select_invalid_db():
         raise Exception('Expect that `%s` does not work' % query)
     except redis.exceptions.ResponseError as ex:
         assert str(ex) == 'invalid DB index'
+
+def test_redis_database_count():
+    """
+    Test enough keyspaces can be prepared if redis_database_count is set to be larger than default 16.
+    Test will fail if redis_database_count isn't set correctly.
+    """
+    logger.info('Make sure redis_database_count is set to be larger than default 16')
+
+    r = connect()
+
+    try:
+        r.execute_command('SELECT 16')  # Use the select to verify the redis_database_count configuration
+        key = random_string(10)
+        r.get(key)
+    except redis.exceptions.ResponseError as ex:
+        if str(ex) == 'invalid DB index':
+            raise Exception('Make sure redis_database_count is set to be larger than default 16')
```

Comment:
So I don't understand, this test will fail if Scylla + redis is just run normally, with the default redis_database_count? It will also fail against regular redis if also run as usual (with a default number of 16)? Why do we want to see this failure all the time?

Reply:
The problem we fixed only occurred when redis_database_count is set to be larger than default 16 for Scylla-redis. The problem is not about verify index in switching db by select command, but not all keyspaces will be created as expected.

So this test will fail if Scylla-redis or regular Redis is just run normally, with the default redis_database_count. It doesn't satisfy our test request.

When we see the error of 'invalid DB index', they will get back to start scylla-redis or regular redis with larger redis_database_count.

Test example:
- start scylla-redis with 20 redis_database_count , or start regular redis with 20 dbs
- select 18  (will success)
- try to write data to database 17 (the idx starts from 0),
    -  expect to succeed
    - scylla-redis will fail before applying this PR.

Reply:
In future, we will start scylla or redis by test code, then we don't need to verify the current redis_database_count.

Reply:
I'm not saying we can't have the tests assume Scylla or Redis was run with specific paramters - we also do that in the Alternator tests, where we assume they were run with a specific setup regarding authentication, otherwise the authentication tests will fail. HOWEVER, if you need to run Scylla and Redis with specific parameters for all the tests to pass, you should document these parameters in redis-test/README.md which explains how to run the tests.
And you need to consider whether complicating the test run for testing this small issue is worth it. Currently the tests don't run Scylla or Redis from scratch, so they don't check **any** of the redis command-line options - it isn't just redis_database_count which they don't check. So maybe it simply shouldn't check any of the command line parameters, and those need to be tested manually (or in the future, with a more general test framework)?

Reply:
The manual test was executed.

There are two options to process the error of database count isn't set to be larger than default.

- option 1: output the err msg and break the subtest, but it still PASSes.
- option 2: raise Exception with err msg and failed the job.

option 2 will reminder us to test manually or test with correct parameter(redis_database_count)

Current test is simple and crude, we can't improve it to be prefect without refactor and add a general test framework.

So I suggest to commit it as current, or remove the automation tests, and add test in future.

Reply:
@nyh ping

Reply:
@amoskong I'll commit for now because you seem to really want this test, but I urge you to rethink this. I think you should do one of the following two options:
1. Document in redis-test/README.md how the user is expected to run Scylla and/or Redis in order for all the tests to pass (we do the same in Alternator, by the way - the user needs to run Scylla in a certain way if he wants the authentication tests to succeed).
2. Mark the tests as "xfail" - meaning that they will fail but this failure is expected, or "skip" - meaning that the test won't even run normally. In either case the user can ask pytest to manually run this test if he wants to try a special configuration with this test.